File: src/devcoach/core/coach.py
"""Rate limit checking and lesson selection logic for devcoach."""

from __future__ import annotations

import sqlite3
from datetime import UTC, datetime, timedelta

from devcoach.core.db import (
    count_filtered_lessons,
    count_lessons_since,
    get_all_knowledge,
    get_knowledge_entries,
    get_knowledge_group_list,
    get_last_lesson_timestamp,
    get_settings,
    get_taught_topic_ids,
    set_feedback,
    upsert_knowledge,
)
from devcoach.core.models import Profile, RateLimitResult


def check_rate_limit(conn: sqlite3.Connection) -> RateLimitResult:
    """Determine whether a new lesson can be delivered.

    Rules (evaluated in order):
    1. Count lessons in the last 24h. If >= max_per_day → denied.
    2. Find timestamp of last lesson. If < min_hours_between ago → denied.
    3. Otherwise: allowed.
    """
    try:
        settings = get_settings(conn)
        now = datetime.now(UTC)

        since_24h = (now - timedelta(hours=24)).isoformat()
        count = count_lessons_since(conn, since_24h)
        if count >= settings.max_per_day:
            return RateLimitResult(
                allowed=False,
                reason=f"Daily limit reached ({count}/{settings.max_per_day} lessons in the last 24h)",
            )

        last_ts = get_last_lesson_timestamp(conn)
        if last_ts is not None:
            last_dt = datetime.fromisoformat(last_ts)
            if last_dt.tzinfo is None:
                last_dt = last_dt.replace(tzinfo=UTC)
            elapsed_minutes = (now - last_dt).total_seconds() / 60
            if elapsed_minutes < settings.min_gap_minutes:
                remaining = settings.min_gap_minutes - elapsed_minutes
                gap_h, gap_m = divmod(settings.min_gap_minutes, 60)
                rem_h, rem_m = divmod(int(remaining), 60)
                return RateLimitResult(
                    allowed=False,
                    reason=(
                        f"Too soon: last lesson {elapsed_minutes:.0f}m ago, "
                        f"minimum interval is {gap_h}h {gap_m}m "
                        f"({rem_h}h {rem_m}m remaining)"
                    ),
                )

        return RateLimitResult(allowed=True)

    except Exception as exc:
        return RateLimitResult(allowed=True, reason=f"Rate limit check failed: {exc}")


def get_profile(conn: sqlite3.Connection) -> Profile:
    """Load and return the current user profile."""
    try:
        return Profile(
            knowledge=get_knowledge_entries(conn),
            groups=get_knowledge_group_list(conn),
        )
    except Exception:
        return Profile(knowledge=[], groups=[])


def apply_knowledge_delta(conn: sqlite3.Connection, topic: str, delta: int) -> Profile:
    """Add delta to the current confidence for a topic (clamped 0-10).

    If the topic does not exist it is created with a base confidence of 5.
    """
    try:
        row = conn.execute("SELECT confidence FROM knowledge WHERE topic = ?", (topic,)).fetchone()
        current = row[0] if row else 5
        upsert_knowledge(conn, topic, current + delta)
        return get_profile(conn)
    except Exception:
        return get_profile(conn)


def record_feedback(
    conn: sqlite3.Connection, lesson_id: str, feedback_value: str | None
) -> str | None:
    """Record feedback for a lesson and auto-adjust knowledge confidence by ±1.

    feedback_value: "know" | "dont_know" | None (to clear).
    Returns the topic_id of the updated lesson, or None if lesson not found.
    """
    topic_id = set_feedback(conn, lesson_id, feedback_value)
    if topic_id and feedback_value in ("know", "dont_know"):
        delta = 1 if feedback_value == "know" else -1
        apply_knowledge_delta(conn, topic_id, delta)
    return topic_id


def get_stats(conn: sqlite3.Connection) -> dict:
    """Return aggregate coaching statistics.

    Returns total_lessons, lessons_today (last 24h), lessons_this_week (last 7d),
    weakest_topics (up to 5), and strongest_topics (up to 5).
    """
    try:
        now = datetime.now(UTC)
        total = count_filtered_lessons(conn)
        today_cutoff = (now - timedelta(hours=24)).isoformat()
        week_cutoff = (now - timedelta(days=7)).isoformat()
        lessons_today = count_lessons_since(conn, today_cutoff)
        lessons_week = count_lessons_since(conn, week_cutoff)
        knowledge = get_all_knowledge(conn)
        sorted_k = sorted(knowledge.items(), key=lambda x: x[1])
        weakest = [{"topic": t, "confidence": c} for t, c in sorted_k[:5]]
        strongest = [{"topic": t, "confidence": c} for t, c in sorted_k[-5:][::-1]]
        return {
            "total_lessons": total,
            "lessons_today": lessons_today,
            "lessons_this_week": lessons_week,
            "weakest_topics": weakest,
            "strongest_topics": strongest,
        }
    except Exception as exc:
        return {"error": str(exc)}


def list_taught_topics(conn: sqlite3.Connection) -> list[str]:
    """Return all topic_ids already present in the lesson log."""
    try:
        return get_taught_topic_ids(conn)
    except Exception:
        return []

File: src/core/coach.ts
// Rate limiting, profile, knowledge deltas, stats, and the lesson-cue decision engine.
import type { DatabaseSync } from "node:sqlite";
import {
  bumpNudge,
  countFilteredLessons,
  countLessonsSince,
  getAllKnowledge,
  getCueState,
  getKnowledgeEntries,
  getKnowledgeGroupList,
  getLastLessonTimestamp,
  getSettings,
  getTaughtTopicIds,
  isOnboardingComplete,
  markCuePending,
  NUDGE_RETRY_AFTER,
  peekNudge,
  setFeedback,
  upsertKnowledge,
} from "./db";
import type { Profile, RateLimitResult } from "./models";

export function checkRateLimit(db: DatabaseSync): RateLimitResult {
  try {
    const settings = getSettings(db);
    const now = new Date();

    const since24h = new Date(now.getTime() - 24 * 3_600_000).toISOString();
    const count = countLessonsSince(db, since24h);
    if (count >= settings.max_per_day) {
      return {
        allowed: false,
        reason: `Daily limit reached (${count}/${settings.max_per_day} lessons in the last 24h)`,
      };
    }

    const lastTs = getLastLessonTimestamp(db);
    if (lastTs != null) {
      const lastDt = new Date(lastTs);
      const elapsedMinutes = (now.getTime() - lastDt.getTime()) / 60000;
      if (elapsedMinutes < settings.min_gap_minutes) {
        const remaining = settings.min_gap_minutes - elapsedMinutes;
        const gapH = Math.floor(settings.min_gap_minutes / 60);
        const gapM = settings.min_gap_minutes % 60;
        const remTrunc = Math.trunc(remaining);
        const remH = Math.floor(remTrunc / 60);
        const remM = remTrunc % 60;
        const agoText =
          elapsedMinutes < 0
            ? `in ${Math.round(-elapsedMinutes)}m (future timestamp)`
            : `${Math.round(elapsedMinutes)}m ago`;
        return {
          allowed: false,
          reason:
            `Too soon: last lesson ${agoText}, minimum interval is ${gapH}h ${gapM}m ` +
            `(${remH}h ${remM}m remaining)`,
        };
      }
    }

    return { allowed: true };
  } catch (err) {
    return { allowed: true, reason: `Rate limit check failed: ${err}` };
  }
}

export function getProfile(db: DatabaseSync): Profile {
  try {
    return { knowledge: getKnowledgeEntries(db), groups: getKnowledgeGroupList(db) };
  } catch {
    return { knowledge: [], groups: [] };
  }
}

export function applyKnowledgeDelta(db: DatabaseSync, topic: string, delta: number): number {
  const current = getAllKnowledge(db)[topic] ?? 5;
  const newConfidence = Math.max(0, Math.min(10, current + delta));
  upsertKnowledge(db, topic, newConfidence);
  return newConfidence;
}

export function recordFeedback(
  db: DatabaseSync,
  lessonId: string,
  feedbackValue: string | null,
): string | null {
  const topicId = setFeedback(db, lessonId, feedbackValue);
  if (topicId && (feedbackValue === "know" || feedbackValue === "dont_know")) {
    applyKnowledgeDelta(db, topicId, feedbackValue === "know" ? 1 : -1);
  }
  return topicId;
}

export function getStats(db: DatabaseSync): Record<string, unknown> {
  try {
    const now = new Date();
    const total = countFilteredLessons(db);
    const todayCutoff = new Date(now.getTime() - 24 * 3_600_000).toISOString();
    const weekCutoff = new Date(now.getTime() - 7 * 86_400_000).toISOString();
    const lessonsToday = countLessonsSince(db, todayCutoff);
    const lessonsWeek = countLessonsSince(db, weekCutoff);
    const knowledge = getAllKnowledge(db);
    const sortedK = Object.entries(knowledge).sort((a, b) => a[1] - b[1]);
    const weakest = sortedK.slice(0, 5).map(([topic, confidence]) => ({ topic, confidence }));
    const strongest = sortedK
      .slice(-5)
      .reverse()
      .map(([topic, confidence]) => ({ topic, confidence }));
    return {
      total_lessons: total,
      lessons_today: lessonsToday,
      lessons_this_week: lessonsWeek,
      weakest_topics: weakest,
      strongest_topics: strongest,
    };
  } catch (err) {
    return { error: String(err) };
  }
}

export function listTaughtTopics(db: DatabaseSync): string[] {
  try {
    return getTaughtTopicIds(db);
  } catch {
    return [];
  }
}

// ── Lesson-cue decision engine (used by the Stop hook, doctor, and prompt-hook) ──

export interface CueDecision {
  cue: boolean;
  /** Human-readable outcome — surfaced by `doctor` and DEVCOACH_HOOK_DEBUG. */
  reason: string;
  nextLessonNumber: number;
}

/**
 * Decide whether this stop cues a lesson. Mutates pacing state: bumps the interaction
 * counter (plan-mode stops never count — planning is not coachable work) and, when it
 * cues, resets the counter and arms the shorter retry threshold (markCuePending).
 * Rate-limited stops keep accumulating so the cue fires at the first allowed stop.
 */
export function evaluateCue(
  db: DatabaseSync,
  sessionId: string | null,
  opts: { planMode: boolean },
): CueDecision {
  const none = (reason: string): CueDecision => ({ cue: false, reason, nextLessonNumber: 0 });
  if (opts.planMode) return none("plan mode (not counted)");
  if (!isOnboardingComplete(db).knowledge_ready) return none("onboarding not complete");

  const settings = getSettings(db);
  let pacing = "pacing disabled (nudge_every=0)";
  if (settings.nudge_every > 0) {
    const n = bumpNudge(db, sessionId ?? "__nosession__", settings.nudge_scope);
    const { pending } = getCueState(db);
    const threshold = pending
      ? Math.min(NUDGE_RETRY_AFTER, settings.nudge_every)
      : settings.nudge_every;
    pacing = `${n}/${threshold}${pending ? " (retry window)" : ""}`;
    if (n < threshold) return none(`paced (${pacing})`);
  }
  const rate = checkRateLimit(db);
  if (!rate.allowed) return none(`rate limited: ${rate.reason ?? "denied"}`);

  markCuePending(db);
  return {
    cue: true,
    reason: `cue (${pacing})`,
    nextLessonNumber: Number(getStats(db).total_lessons ?? 0) + 1,
  };
}

/**
 * Read-only dry run of `evaluateCue` for the NEXT eligible stop — no bump, no state
 * change. Drives `doctor`'s verdict and the prompt-hook's priming decision.
 */
export function explainCue(
  db: DatabaseSync,
  sessionId: string | null,
): { wouldCue: boolean; reasons: string[] } {
  try {
    const reasons: string[] = [];
    let wouldCue = true;
    if (!isOnboardingComplete(db).knowledge_ready) {
      return { wouldCue: false, reasons: ["onboarding not complete — no knowledge topics yet"] };
    }
    const settings = getSettings(db);
    if (settings.nudge_every > 0) {
      const n = peekNudge(db, sessionId ?? "__nosession__", settings.nudge_scope) + 1;
      const { pending } = getCueState(db);
      const threshold = pending
        ? Math.min(NUDGE_RETRY_AFTER, settings.nudge_every)
        : settings.nudge_every;
      if (n < threshold) {
        wouldCue = false;
        reasons.push(
          `pacing: next stop would be ${n}/${threshold}${pending ? " (retry window)" : ""}`,
        );
      } else {
        reasons.push(`pacing: next stop reaches the threshold (${n}/${threshold})`);
      }
    } else {
      reasons.push("pacing disabled (nudge_every=0) — every eligible stop cues");
    }
    const rate = checkRateLimit(db);
    if (rate.allowed) {
      reasons.push("rate limit: allowed");
    } else {
      wouldCue = false;
      reasons.push(`rate limited: ${rate.reason ?? "denied"}`);
    }
    return { wouldCue, reasons };
  } catch (err) {
    return { wouldCue: false, reasons: [`check failed: ${err}`] };
  }
}
